src/helpers/util.js

```diff
@@ -39,5 +39,32 @@ module.exports = {
       ret.push(row)
     }
     return ret
+  },
+
+  /*
+    binary search
+  */
+  findLowerBound: function (target, changes) {
+    if (changes.length === 0) {
+      return undefined
+    }
+
+    if (changes.length === 1) {
+      if (changes[0] > target) {
+        // we only a closest maximum, returning O
+        return 0
+      } else {
+        return changes[0]
+      }
+    }
+
+    var middle = Math.floor(changes.length / 2)
+    if (changes[middle] > target) {
+      return this.findLowerBound(target, changes.slice(0, middle))
```

Comment:
I think we should rather use a second `findLowerBoundInternal` function and indices than `slice`.